### PR TITLE
Cache log implementation

### PR DESCRIPTION
**Pull request by bkolad**

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    pub fn get_value(&self, key: &CacheKey) -> ExistsInCache {
+        match self.log.get(key) {
+            Some(value) => ExistsInCache::Yes(value.last_value().clone()),
+            None => ExistsInCache::No,
+        }
+    }
+
+    /// The first read for a given key is inserted in the cache. For an existing cache entry
+    /// checks if reads are consistent with previous reads/writes.
+    pub fn add_read(&mut self, item: CacheEntry) {
```
Comment (preston-evans98) [inline src/cache.rs:46]: non-blocking: I'm curious why the argument to this function is a struct instead of:
```
pub fn add_read(&mut self, key: CacheKey, value: Option<CacheValue>)
```

File: src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        fn get_value(self) -> CacheEntry {
+            match self {
+                ReadWrite::Read(r) => r,
+                ReadWrite::Write(w) => w,
+            }
+        }
+
+        fn check_cache_consistency(self, rhs: Self, merged: &CacheLog) {
+            match (self, rhs) {
+                (ReadWrite::Read(r_l), ReadWrite::Read(r_r)) => {
```
Comment (preston-evans98) [inline src/cache.rs:170]: nit: `r_l` and `r_r` are hard to parse. Should we expand to `read_l` and `read_r` or even `read_left` and `read_right`?

File: src/lib.rs [hunk 3]
```diff
@@ -1,14 +1,10 @@
...
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
-    }
-}
+use std::sync::Arc;
+
+// Arc here is used for cheap cloning.
+pub type CacheKey = Arc<Vec<u8>>;
+pub type CacheValue = Arc<Vec<u8>>;
```
Comment (preston-evans98) [inline src/lib.rs:10]: non-blocking: using two type aliases to `Arc<Vec<u8>>` here means that the type checker won't catch key-value confusion. We should consider switching to incompatible types or using the newtype pattern

File: src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+                    right_cache.add_to_cache(right_rw);
+                }
+            }
+
+            let result = std::panic::catch_unwind(|| {
+                left_cache.merge(right_cache);
+            });
+
+            // Assert that merge failed
+            assert!(result.is_err());
```
Comment (preston-evans98) [inline src/cache.rs:312]: nit: We can replace all of this with a `#[should_panic]` annotation at the top of the test.

**Review by preston-evans98 — CHANGES_REQUESTED** [review]

File: src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    No,
+}
+
+/// CacheLog keeps track of the first write and the last read for a given key.
+#[derive(Default)]
+pub struct CacheLog {
+    log: HashMap<CacheKey, Access>,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
```
Comment (preston-evans98) [inline src/cache.rs:23]: Missing doc comment:
```suggestion
An entry to be added to the cache, consisting of a `key` and an optional `value`. If the value is `None`, writing the entry will be treated as a deletion.
#[derive(PartialEq, Eq, Clone, Debug)]
```

File: src/access.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            Access::ReadThenWrite { modified, .. } => *modified = new_value,
+            // For Write override the original value with a new value
+            Access::Write(value) => *value = new_value,
+        }
+    }
+
+    pub(crate) fn merge(self, rhs: Self) -> Self {
+        match (self, rhs) {
+            (Access::Read(l), Access::Read(r)) => {
+                assert_eq!(l, r);
```
[the remark this answers is not in the document]
Comment (bkolad) [reply src/access.rs]: done

File: src/cache.rs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        }
+    }
+
+    /// The first read for a given key is inserted in the cache. For an existing cache entry
+    /// checks if reads are consistent with previous reads/writes.
+    pub fn add_read(&mut self, item: CacheEntry) {
+        match self.log.entry(item.key) {
+            Entry::Occupied(existing) => {
+                let last_value = existing.get().last_value();
+                assert_eq!(last_value, &item.value);
```
[the remark this answers is not in the document]
Comment (bkolad) [reply src/cache.rs]: done

File: src/cache.rs (hunk 4 above)
Comment (bkolad) [reply src/cache.rs]: fixed

File: src/lib.rs (hunk 3 above)
Comment (bkolad) [reply src/lib.rs]: done

File: src/access.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        assert_eq!(
+            left,
+            Access::ReadThenWrite {
+                original: create_value(first_read),
+                modified: create_value(last_write)
+            }
+        )
+    }
```
[the remark this answers is not in the document]
Comment (bkolad) [reply src/access.rs:246]: Done #4

File: src/cache.rs [hunk 9]
```diff
@@ -20,14 +30,24 @@ pub struct CacheLog {
...
+impl CacheLog {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            log: HashMap::with_capacity(capacity),
+        }
+    }
+}
+
+/// An entry to be added to the cache, consisting of a `key` and an optional `value`.
+/// If the value is `None`, writing the entry will be treated as a deletion.
```
Comment (preston-evans98) [inline src/cache.rs:42]: non-blocking: Doc comment is out of date now :)

File: src/cache.rs (hunk 2 above)
Comment (bkolad) [reply src/cache.rs:200]: done

File: src/cache.rs (hunk 5 above)
Comment (bkolad) [reply src/cache.rs:43]: merged

File: src/cache.rs (hunk 1 above)
Comment (bkolad) [reply src/cache.rs]: done: I have changed the signature.

File: src/cache.rs (hunk 9 above)
Comment (bkolad) [reply src/cache.rs:42]: After changing the `add_read & add_write` signature the `CacheEntry` is used only in tests so I moved it there.

File: src/cache.rs (hunk 2 above)
Comment (bkolad) [reply src/cache.rs:200]: done

**Review by preston-evans98 — APPROVED** [review]
